### ISSUE TITLE
Ruby + Bundler in Sandbox

### ISSUE THREAD

**Issue by tbeseda** [architect/architect#1291]
## Describe the issue

Running a Ruby function locally in Sandbox fails when using bundled gems.  
I may be doing this wrong, but I expected to not need to require `'bundler/setup'` -- the docs do not indicate this, and seem to say that Sandbox works with Ruby + Bundler automagically.

## Local env

I'm using a new M1 Mac with Monterey + rbenv.

|   | version |
|---|---|
| Node | 14.18.2 |
| npm | 7.24.2 |
| Ruby | 2.7.5p203 |
| gem | 3.1.6 |
| bundler | 2.2.33 |
| Architect | 9.4.2 |
| Sandbox | 4.4.2 |

## Steps to reproduce

1. `npm init @architect arc-ruby`
2. replace `get-index/index.js` with a simple `index.rb`:

```rb
require 'json'

def handler
  { body: 'hello from Ruby!' }
end
```
3. add a `config.arc` next to `index.rb`:

```
@aws
runtime ruby
```
4. run `npm start` and visit localhost:3333 -- works 👍 
5. inside the `get-index/` function folder run:

```sh
bundle init
bundle config set --local path 'vendor/bundle'
# ^-- this is the preferred way instead of "bundle install --path vendor/bundle"
bundle add architect-functions
```
This works; I now have a Gemfile like
```gemfile
source "https://rubygems.org"
git_source(:github) { |repo_name| "https://github.com/#{repo_name}" }
gem "architect-functions", "~> 0.5.3"
```
and a big bag of gems in `./src/http/get-index/vendor/bundle/ruby/2.7.0/gems`

6. add `architect/functions` to index.rb:

```diff
+ require 'architect/functions`
require 'json'

def handler
  { body: 'hello from Ruby!' }
end
```

7. from the project root: `npm start` and visit localhost:3333:

```sh
Error

Process exited with 1

/Users/tbeseda/.rbenv/versions/2.7.5/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:83:in `require': cannot load such file -- architect/functions (LoadError)
	from /Users/tbeseda/.rbenv/versions/2.7.5/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:83:in `require'
	from /Users/tbeseda/dev/scratch/arc-ruby/src/http/get-index/index.rb:4:in `'
	from /Users/tbeseda/.rbenv/versions/2.7.5/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:83:in `require'
	from /Users/tbeseda/.rbenv/versions/2.7.5/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:83:in `require'
	from -e:8:in `
'
```
😧 

8. add to index.rb
 
```diff
+ require 'bundler/setup`
require 'architect/functions`
require 'json'

def handler
  { body: 'hello from Ruby!' }
end
```

9. visit localhost:3333 and it works 👍 

If I've got the process incorrect, we should update the docs. I'm happy to work on that, just need to know the canonical way to run Ruby with Sandbox. 

Thanks!

**Comment by tbeseda** [comment]
<details>

<summary>gem environment<summary>

```sh
❯ gem environment
RubyGems Environment:
  - RUBYGEMS VERSION: 3.1.6
  - RUBY VERSION: 2.7.5 (2021-11-24 patchlevel 203) [arm64-darwin21]
  - INSTALLATION DIRECTORY: /Users/tbeseda/.rbenv/versions/2.7.5/lib/ruby/gems/2.7.0
  - USER INSTALLATION DIRECTORY: /Users/tbeseda/.gem/ruby/2.7.0
  - RUBY EXECUTABLE: /Users/tbeseda/.rbenv/versions/2.7.5/bin/ruby
  - GIT EXECUTABLE: /usr/bin/git
  - EXECUTABLE DIRECTORY: /Users/tbeseda/.rbenv/versions/2.7.5/bin
  - SPEC CACHE DIRECTORY: /Users/tbeseda/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /Users/tbeseda/.rbenv/versions/2.7.5/etc
  - RUBYGEMS PLATFORMS:
    - ruby
    - arm64-darwin-21
  - GEM PATHS:
     - /Users/tbeseda/.rbenv/versions/2.7.5/lib/ruby/gems/2.7.0
     - /Users/tbeseda/.gem/ruby/2.7.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - redacted
```

</details>

**Comment by ryanblock** [comment]
That seems fine, but tbh I don't know what the correct Rubyist way of doing this would be. @brianleroux?

Semi-related: this issue is 💯 and I will be sending it along to everyone in the future wondering how to write a good issue. 🌈

**Comment by brianleroux** [comment]
ok! that was a bit of a fun adventure for me. 'fun'. haha

anyhow, learned some new things:
- we def need `require 'bundler/setup'` to be added to the docs
- for deployment Lambda bundler also wants `bundle lock --add-platform x86_64-linux`
- `arc deploy` is broken; but `arc deploy --no-hydrate` works; we should recommend that for now

bundler need write permissions (why?!) but Lambda is readonly; to fix the dev needs to do one of these things:
- set an env var `are env staging BUNDLE_FROZEN 1` (and again for prod `arc env production BUNDLE_FROZEN 1`
- or, locally, in the function folder run `bundle config set deployment true` 

also fixed a bug in `architect/create` and `architect/sandbox` around the required keyword arguments; this lead me to filing https://github.com/architect/architect/issues/1293 which could be a fun one to bang out.

**Comment by tbeseda** [comment]
Thanks for the research here @brianleroux 

I took some direction from your findings and created an Arc project with a few endpoints to demonstrate some Ruby basics and a couple with Bundler onboard: https://github.com/tbeseda/test-arc-ruby0
I'm adding some more info about my findings to the Readme over there, but everything you mentioned tracks and does help get it deployed.
Happy to demo it live as well.

I also updated our docs with some recommendations for deploying with Bundler https://github.com/architect/arc.codes/pull/479

Looks like there some Sandbox and architect/function-ruby work that needs to be done to catch up to Lambda's provided environment

IMO this can be broken down into a few smaller issues, so I'll close this one. Definitely open to helping out and/or pairing on this. Even if my Ruby is super rusty :)
